src/object.jl

```diff
@@ -372,3 +372,135 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
+
+Data chunks are written to a channel and spawned tasks read buffers from this channel.
+Each task uploads a distinct part to blob storage.
+Each part is tagged with an id and we keep the tags into a vector. We also increment a
+counter to keep track of the number of parts. When there are no more data chunks to upload,
+we send a POST request with a single id for the entire upload.
+
+# Arguments
+* `store::AbstractStore`: The S3 Bucket / Azure Container object
+* `key::String`: S3 key / Azure blob resource name
+
+# Keywords
+* `credentials::Union{CloudCredentials, Nothing}=nothing`: Credentials object used in HTTP
+    requests
+* `kwargs...`: HTTP keyword arguments are forwarded to underlying HTTP requests,
+
+## Examples
+```
+# Get an IO stream for a remote CSV file `test.csv` living in your S3 bucket
+io = MultipartUploadStream(bucket, "test.csv"; credentials)
+```
+"""
+mutable struct MultipartUploadStream <: IO
+    store::AbstractStore
+    url::String
+    credentials::Union{Nothing, AWS.Credentials, Azure.Credentials}
+    uploadState
+    eTags::Vector{String}
+    upload_queue::Channel{Tuple{Int, Vector{UInt8}}}
+    sync::OrderedSynchronizer
+    cond_wait::Threads.Condition
+    cur_part_id::Int
+    ntasks::Int
+    is_error::Bool
+    exc::Union{Exception, Nothing}
+    sem::Base.Semaphore
+
+    function MultipartUploadStream(
+        store::AbstractStore,
+        key::String;
+        credentials::Union{CloudCredentials, Nothing}=nothing,
+        concurrent_writes_to_channel::Int=(4 * Threads.nthreads()),
+        kw...
+    )
+        url = makeURL(store, key)
+        uploadState = API.startMultipartUpload(store, key; credentials, kw...)
+        io = new(
+            store,
+            url,
+            credentials,
+            uploadState,
+            Vector{String}(),
+            Channel{Tuple{Int, Vector{UInt8}}}(Inf),
+            OrderedSynchronizer(1),
+            Threads.Condition(),
+            0,
+            0,
+            false,
+            nothing,
+            Base.Semaphore(concurrent_writes_to_channel)
+        )
+        return io
+    end
+end
+
+function Base.write(io::MultipartUploadStream, bytes::Vector{UInt8}; kw...)
+    local part_n
+    Base.@lock io.cond_wait begin
+        io.ntasks += 1
+        io.cur_part_id += 1
+        part_n = io.cur_part_id
+        notify(io.cond_wait)
+    end
+    Base.acquire(io.sem)
+    put!(io.upload_queue, (part_n, bytes))
+    Base.release(io.sem)
+    Threads.@spawn _upload_task($io; $(kw)...)
+    return nothing
+end
+
+function Base.wait(io::MultipartUploadStream)
+    try
+        Base.@lock io.cond_wait begin
+            while true
+                io.ntasks == 0 && !io.is_error && break
+                if io.is_error
+                    throw(io.exc)
+                    break
+                end
+                wait(io.cond_wait)
+            end
+        end
+    catch e
+        rethrow()
```

Comment:
I think we are missing a way to signal to S3/Blobs that we're aborting the multi-part upload a la [AbortMultipartUpload](`https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html`)... can you open an issue about it?

Also, I think we should consider the following syntax to the user

```julia
MultipartUploadStream(...) do io 
    ...
end
```

something like
```julia
function MultipartUploadStream(f::Function, args...; kwargs...)
    io = MultipartUploadStream(args...; kwargs...)
    try
        f(io)
        wait(io)
        close(io)
    catch e
        abort(io, e) # todo, we don't have this yet
        rethrow()
    end
end
```

Reply:
There is already an issue for adding more low-level functionality regarding multipart uploads including list parts and and list parts that are in progress, which are related to aborting: https://github.com/JuliaServices/CloudStore.jl/issues/3. Should I add a comment there about abort?

Reply:
For the alternative syntax, `f` would need to encapsulate a call to `MultipartUploadStream.write()` though, right?